Firestore/Source/Core/FSTSyncEngine.mm

```diff
@@ -491,42 +490,41 @@ - (void)updateTrackedLimboDocumentsWithChanges:(NSArray<FSTLimboDocumentChange *
 }
 
 - (void)trackLimboChange:(FSTLimboDocumentChange *)limboChange {
-  FSTDocumentKey *key = limboChange.key;
+  DocumentKey key{limboChange.key};
 
-  if (!self.limboTargetsByKey[key]) {
-    FSTLog(@"New document in limbo: %@", key);
-    FSTTargetID limboTargetID = _targetIdGenerator.NextId();
-    FSTQuery *query = [FSTQuery queryWithPath:key.path];
+  if (_limboTargetsByKey.find(key) == _limboTargetsByKey.end()) {
+    FSTLog(@"New document in limbo: %s", key.ToString().c_str());
+    TargetId limboTargetID = _targetIdGenerator.NextId();
+    FSTQuery *query = [FSTQuery queryWithPath:key.path()];
     FSTQueryData *queryData = [[FSTQueryData alloc] initWithQuery:query
                                                          targetID:limboTargetID
                                              listenSequenceNumber:kIrrelevantSequenceNumber
                                                           purpose:FSTQueryPurposeLimboResolution];
-    self.limboKeysByTarget[@(limboTargetID)] = key;
+    _limboKeysByTarget[limboTargetID] = key;
     [self.remoteStore listenToTargetWithQueryData:queryData];
-    self.limboTargetsByKey[key] = @(limboTargetID);
+    _limboTargetsByKey[key] = limboTargetID;
   }
 }
 
 /** Garbage collect the limbo documents that we no longer need to track. */
 - (void)garbageCollectLimboDocuments {
   const std::set<DocumentKey> garbage = [self.limboCollector collectGarbage];
   for (const DocumentKey &key : garbage) {
-    FSTBoxedTargetID *limboTarget = self.limboTargetsByKey[static_cast<FSTDocumentKey *>(key)];
-    if (!limboTarget) {
+    const auto iter = _limboTargetsByKey.find(key);
+    if (iter == _limboTargetsByKey.end()) {
       // This target already got removed, because the query failed.
       return;
     }
-    FSTTargetID limboTargetID = limboTarget.intValue;
+    TargetId limboTargetID = iter->second;
     [self.remoteStore stopListeningToTargetID:limboTargetID];
-    [self.limboTargetsByKey removeObjectForKey:key];
-    [self.limboKeysByTarget removeObjectForKey:limboTarget];
+    _limboTargetsByKey.erase(key);
+    _limboKeysByTarget.erase(limboTargetID);
   }
 }
 
 // Used for testing
-- (NSDictionary<FSTDocumentKey *, FSTBoxedTargetID *> *)currentLimboDocuments {
-  // Return defensive copy
-  return [self.limboTargetsByKey copy];
+- (const std::map<DocumentKey, TargetId> &)currentLimboDocuments {
+  return _limboTargetsByKey;
```

Comment:
You've skipped creating the copy. I don't have any context into why a copy was returned previously; but judging from the comment, it looks like it was just for safety in a method whose performance is irrelevant. If it doesn't hurt anything, keeping the copy might be a good idea.

(If keeping the copy still works, but you feel the copy shouldn't occur, a separate CL might be a better venue for that, as it's independent of this change.)

Reply:
Done. However, could elaborate more on this? The only caller already made a copy:

Firestore/Example/Tests/SpecTests/FSTSpecTests.mm:  std::map<DocumentKey, TargetId> actualLimboDocs = self.driver.currentLimboDocuments;

Do we need to copy twice? (The old objc code copied twice, once here as `return [self.limboTargetsByKey copy];` and once in the test as `NSMutableDictionary<FSTDocumentKey *, FSTBoxedTargetID *> *actualLimboDocs =
      [NSMutableDictionary dictionaryWithDictionary:self.driver.currentLimboDocuments];`

Reply:
We don't need to copy twice, and return value optimization ensures we don't.

However, it's important that the contract exposed here is a copy and not a reference to the underlying data. This prevents the caller from just using the reference without making a copy.